packages/engine/Source/Scene/Megatexture.js

```diff
@@ -389,130 +321,173 @@ Megatexture.prototype.isFull = function () {
-      break;
-    } else {
-      textureDimension *= 2;
+  // Find a nearby number with no prime factor larger than 7.
+  const factors = findFactorsOfNearbyComposite(tileCount, maxTileCount);
+
+  // Split these factors into the three dimensions, keeping the total texture size
+  // smaller than the maximum in each dimension.
+  for (let i = 0; i < 3; i++) {
+    const [axis, length] = sortedDimensions[i];
+    const maxTileCountAlongAxis = Math.floor(maximum3DTextureSize / length);
+    const axisFactors = getDimensionFromFactors(factors, maxTileCountAlongAxis);
+    textureDimension[axis] = getProductOfFactors(axisFactors) * length;
+    // Remove used factors.
+    for (let j = 0; j < factors.length; j++) {
+      factors[j] -= axisFactors[j];
     }
   }
 
-  const textureMemoryByteLength =
-    textureDimension * textureDimension * channelCount * datatypeSizeInBytes;
-  return textureMemoryByteLength;
+  return textureDimension;
 };
 
 /**
- * Write an array of tile metadata to the megatexture.
- * @param {number} index The index of the tile's location in the megatexture.
- * @param {Float32Array|Uint16Array|Uint8Array} data The data to be written.
+ * Approximate an integer by a product of small prime factors (2, 3, 5, and 7).
+ * @private
+ * @param {number} n The integer to be approximated
+ * @param {number} maxN A maximum integer which the approximation should not exceed
+ * @returns {number[]} The exponents of the prime factors 2, 3, 5, and 7 whose product is the approximation of n
  */
-Megatexture.prototype.writeDataToTexture = function (index, data) {
-  // Unsigned short textures not allowed in webgl 1, so treat as float
-  const tileData =
-    data.constructor === Uint16Array ? new Float32Array(data) : data;
-
-  const {
-    tileVoxelDataTemp,
-    voxelCountPerTile,
-    sliceCountPerRegion,
-    voxelCountPerRegion,
-    channelCount,
-    regionCountPerMegatexture,
-  } = this;
-
-  for (let z = 0; z < voxelCountPerTile.z; z++) {
-    const sliceVoxelOffsetX = (z % sliceCountPerRegion.x) * voxelCountPerTile.x;
-    const sliceVoxelOffsetY =
-      Math.floor(z / sliceCountPerRegion.x) * voxelCountPerTile.y;
-    for (let y = 0; y < voxelCountPerTile.y; y++) {
-      const readOffset = getReadOffset(voxelCountPerTile, y, z);
-      const writeOffset =
-        (sliceVoxelOffsetY + y) * voxelCountPerRegion.x + sliceVoxelOffsetX;
-      for (let x = 0; x < voxelCountPerTile.x; x++) {
-        const readIndex = readOffset + x;
-        const writeIndex = writeOffset + x;
-        for (let c = 0; c < channelCount; c++) {
-          tileVoxelDataTemp[writeIndex * channelCount + c] =
-            tileData[readIndex * channelCount + c];
-        }
-      }
-    }
+function findFactorsOfNearbyComposite(n, maxN) {
+  n = Math.min(n, maxN);
+  if (Math.floor(n) !== n) {
+    throw new DeveloperError("n and maxN must be integers");
+  } else if (n < 1) {
+    throw new DeveloperError("n and maxN must be at least 1");
+  }
```

Comment:
Would it make sense to gate these checks behind a `//>>includeStart('debug', pragmas.debug);`? Asking more for my own understanding of CesiumJS best practice than an actual preference. :)